src/core/manager.c

```diff
@@ -3665,8 +3665,70 @@ static int manager_run_environment_generators(Manager *m) {
         return r;
 }
 
+static int build_generator_environment(Manager *m, char ***ret) {
+        _cleanup_strv_free_ char **nl = NULL;
+        Virtualization v;
+        int r;
+
+        assert(m);
+        assert(ret);
+
+        /* Generators oftentimes want to know some basic facts about the environment they run in, in order to
+         * adjust generated units to that. Let's pass down some bits of information that are easy for us to
+         * determine (but a bit harder for generator scripts to determine), as environment variables. */
+
+        nl = strv_copy(m->transient_environment);
+        if (!nl)
+                return -ENOMEM;
+
+        r = strv_env_assign(&nl, "SYSTEMD_SCOPE", MANAGER_IS_SYSTEM(m) ? "system" : "user");
+        if (r < 0)
+                return r;
+
+        if (MANAGER_IS_SYSTEM(m)) {
+                /* Note that $SYSTEMD_INITRD may be used to override the initrd detection in much of our
+                 * codebase. This is hence more than purely informational. It will shorcut detection of the initrd
+                 * state if generators invoke our own tools. But that's OK, as it would come to the same results
+                 * (hopefully) */
+                r = strv_env_assign(&nl, "SYSTEMD_IN_INITRD", one_zero(in_initrd()));
+                if (r < 0)
+                        return r;
+
+                if (m->first_boot >= 0) {
+                        r = strv_env_assign(&nl, "SYSTEMD_FIRST_BOOT", one_zero(m->first_boot));
+                        if (r < 0)
+                                return r;
+                }
+        }
+
+        v = detect_virtualization();
+        if (v < 0)
+                log_debug_errno(v, "Failed to detect virtualization, ignoring: %m");
+        else if (v > 0) {
+                r = strv_env_assign(&nl, "SYSTEMD_VIRTUALIZATION", virtualization_to_string(v));
+                if (r < 0)
+                        return r;
+
+                r = strv_env_assign(&nl, "SYSTEMD_VIRTUALIZATION_IS_VM", one_zero(VIRTUALIZATION_IS_VM(v)));
+                if (r < 0)
+                        return r;
+
+                r = strv_env_assign(&nl, "SYSTEMD_VIRTUALIZATION_IS_CONTAINER", one_zero(VIRTUALIZATION_IS_CONTAINER(v)));
+                if (r < 0)
+                        return r;
+        }
```

Comment:
Hmm, let's maybe define this a bit differently.
SYSTEMD_VIRTUALIZATION="type[:detail]", where "type" would be one of "", "chroot", "container", "vm", and the second part would only be set if known, for "vm" and "container" types. This way this becomes generic and the same variable can be used for other purposes later without too much verbosity.

Reply:
In particular, if we want to allow overriding autodetection with a variable in the future, it's much easier to have just one variable, so that there can be no conflicting information.

Reply:
hmm, my assumption was that it is much easier from a shell script to parse $SYSTEMD_VIRTUALIZATION_IS_CONTAINER, and that's what 99% of the generators probably care about.

i for one wouldn't know how to test the var quickly if for "is this a container?" if i get "container:systemd-nspawn"?

Reply:
In C: `startswith_ptr(getenv("SYSTEMD_VIRTUALIZATION"), "container")`, in bash: `[[ $x = *container* ]]`, in Python `os.getenv('SYSTEMD_VIRTUALIZATION', '').startswith('container')`, etc. I don't think it's any harder to check.

Reply:
and in bourne shell?

Reply:
I must say, I disagree with this request. If I was to write a generator script in shell I'd probably just shell out `systemd-detect-virt -c` instead of trying to parse the `container:` thing out of the env var... But anyway, I reworked this as requested now, given I am not the audience for this feature anyway...

Reply:
If it turns out to be a problem, you have the right to say blame it on me ;)